presto-cassandra/src/main/java/com/facebook/presto/cassandra/CassandraTokenSplitManager.java

```diff
@@ -13,129 +13,132 @@
+    public CassandraTokenSplitManager(CassandraSession session, CassandraClientConfig config)
     {
-        this.cassandraThriftClient = new CassandraThriftClient(requireNonNull(connectionFactory, "connectionFactory is null"));
-        this.executor = requireNonNull(executor, "executor is null");
-        this.splitSize = config.getSplitSize();
-        try {
-            this.partitioner = FBUtilities.newPartitioner(config.getPartitioner());
-        }
-        catch (ConfigurationException e) {
-            throw new RuntimeException(e);
-        }
+        this(session, config.getSplitSize());
     }
 
-    public List<TokenSplit> getSplits(String keyspace, String columnFamily)
-            throws IOException
+    public CassandraTokenSplitManager(CassandraSession session, int splitSize)
     {
-        List<TokenRange> masterRangeNodes = cassandraThriftClient.getRangeMap(keyspace);
-
-        // canonical ranges, split into pieces, fetching the splits in parallel
-        List<TokenSplit> splits = new ArrayList<>();
-        List<Future<List<TokenSplit>>> splitFutures = new ArrayList<>();
-        for (TokenRange range : masterRangeNodes) {
-            // for each range, pick a live owner and ask it to compute bite-sized splits
-            splitFutures.add(executor.submit(new SplitCallable<>(range, keyspace, columnFamily, splitSize, cassandraThriftClient, partitioner)));
-        }
+        this.session = requireNonNull(session, "session is null");
+        this.splitSize = splitSize;
+    }
 
-        // wait until we have all the results back
-        for (Future<List<TokenSplit>> futureInputSplits : splitFutures) {
-            try {
-                splits.addAll(futureInputSplits.get());
+    public List<TokenSplit> getSplits(String keyspace, String table)
+            throws IOException
+    {
+        Set<TokenRange> tokenRanges = getTokenRanges();
+        List<SizeEstimate> estimates = getSizeEstimates(keyspace, table);
+        Map<TokenRange, SizeEstimate> estimateMap = mapSizeEstimates(estimates);
+
+        ImmutableList.Builder<TokenSplit> splits = ImmutableList.builder();
+
+        for (TokenRange tokenRange : tokenRanges) {
+            List<String> endpoints = unmodifiableList(getReplicas(keyspace, tokenRange)
+                    .stream()
+                    .map(Host::toString)
+                    .collect(toList()));
+
+            SizeEstimate sizeEstimate = estimateMap.get(tokenRange);
+            // TODO: split range based on the partition size, not on the partition count
+            long partitionsCount = 0;
+            if (sizeEstimate != null) {
+                partitionsCount = sizeEstimate.getPartitionsCount();
             }
-            catch (Exception e) {
-                throw new IOException("Could not get input splits", e);
+            int subRangesCount = toIntExact((partitionsCount / splitSize) + 1);
+            List<TokenRange> subRanges = tokenRange.splitEvenly(subRangesCount);
```

Comment:
CassandraThriftClient#getSubSplits does have a fallback --
``` 
catch (TApplicationException e) {
  if (e.getType() == TApplicationException.UNKNOWN_METHOD) {
  List<String> splitPoints = client.describe_splits(columnFamily, range.start_token, range.end_token, splitSize);
  return tokenListToSplits(splitPoints, splitSize);
 }
}
```
We should do something similar so that there isn't a regression.